src/Microsoft.OData.Client/ALinq/ResourceSetExpression.cs

```diff
@@ -104,11 +104,13 @@ internal override QueryableResourceExpression CreateClone()
         /// <returns>A copy of this with the new types</returns>
         protected override QueryableResourceExpression CreateCloneWithNewTypes(Type newType, Type newResourceType)
         {
+            bool ignoreNewResourceType = newResourceType.IsAbstract || newResourceType.IsInterface;
+            
             return new ResourceSetExpression(
                 newType,
                 this.source,
                 this.MemberExpression,
-                newResourceType,
+                ignoreNewResourceType ? this.ResourceType : newResourceType,

```

Comment:
What is the impact of this change? Is this something safe to do?

Reply:
This change was made in order to address what I mention here: https://github.com/OData/odata.net/issues/2856#issuecomment-1960616567

I cannot imagine any scenario where this change could be a problem. Also I have been intensively using my modified fork for 3 months, every single day, and this caused 0 issues. So I would say It's safe.

Reply:
You mean this issue happens when you add OriginalName attribute to the interface properties then use OrderBy?

Reply:
It happens even without the OriginalName attribute. It's just a casting side effect.

Here is a minimalist example to reproduce the issue:

```cs

using Microsoft.OData.Client;


Uri uri = new Uri("https://services.odata.org/V4/Northwind/Northwind.svc");
DataServiceContext ctx = new DataServiceContext(uri);        
DataServiceQuery<CustomerClass> originalQuery = ctx.CreateQuery<CustomerClass>("Customers");

IQueryable<ICustomer> castedQuery = originalQuery;
castedQuery = castedQuery.OrderBy(c => c.CustomerID).Take(1);

// This will throw the mentioned exception at materialization level because it cannot instance an ICustomer
var result = castedQuery.ToList();


// Types declarations:

public interface ICustomer
{
    string CustomerID { get; set; }
}

[Key("CustomerID")]
public class CustomerClass : ICustomer
{
    public string CustomerID { get; set; }
}

```